src/Voron/Impl/LowLevelTransaction.cs

```diff
@@ -663,91 +487,73 @@ private Page GetPageInternal(long pageNumber)
 
-            DataPager.TryReleasePage(this, pageNumber);
+            DataPager.TryReleasePage(ref PagerTransactionState, pageNumber);
         }
 
         private void ThrowObjectDisposed()
         {
-            if (_txState.HasFlag(TxState.Disposed))
+            if (_txStatus.HasFlag(TxStatus.Disposed))
                 throw new ObjectDisposedException("Transaction is already disposed");
 
-            if (_txState.HasFlag(TxState.Errored))
+            if (_txStatus.HasFlag(TxStatus.Errored))
                 throw new InvalidDataException("The transaction is in error state, and cannot be used further");
 
-            throw new ObjectDisposedException("Transaction state is invalid: " + _txState);
+            throw new ObjectDisposedException("Transaction state is invalid: " + _txStatus);
         }
 
         public Page AllocatePage(int numberOfPages, long? pageNumber = null, Page? previousPage = null, bool zeroPage = true)
         {
-            if (pageNumber == null)
+            pageNumber ??= GetPageNumberForAllocation(numberOfPages);
+            return AllocatePageImpl(numberOfPages, pageNumber.Value, previousPage, zeroPage);
+        }
+        
+        public Page AllocateMultiplePageAndReturnFirst(int numberOfPages, bool zeroPage = true)
+        {
+            var firstPageNumber = GetPageNumberForAllocation(numberOfPages);
+            var first = AllocatePageImpl(1, firstPageNumber, previousVersion: null, zeroPage);
+            for (int i = 1; i < numberOfPages; i++)
             {
-                EnsureNotCurrentlyHoldingRootObjectsOpen();
-                pageNumber = _freeSpaceHandling.TryAllocateFromFreeSpace(this, numberOfPages);
-                if (pageNumber == null) // allocate from end of file
-                {
-                    pageNumber = State.NextPageNumber;
-                    State.UpdateNextPage(State.NextPageNumber + numberOfPages);
-                }
+                AllocatePageImpl(1, firstPageNumber + i, previousVersion: null, zeroPage);
             }
-            return AllocatePage(numberOfPages, pageNumber.Value, previousPage, zeroPage);
+            return first;
+        }
+
+        private long GetPageNumberForAllocation(int numberOfPages)
+        {
+            EnsureNotCurrentlyHoldingRootObjectsOpen();
+            var pageNumber = _freeSpaceHandling.TryAllocateFromFreeSpace(this, numberOfPages);
+            if (pageNumber != null)
+                return pageNumber.Value;
+
+            // allocate from end of file
+            var eof = _envRecord.NextPageNumber + _localTxNextPageNumber;
+            _localTxNextPageNumber += numberOfPages;
+            return eof;
+        }
+
+        public long GetNextPageNumber() => _envRecord.NextPageNumber + _localTxNextPageNumber;
+
+        public void SetNextPageNumber(long nextPageNumber) => _localTxNextPageNumber = nextPageNumber - _envRecord.NextPageNumber;
+
+        internal void TestOnly_SetLocalTxNextPageNumber(long n)
```

Comment:
Can we make a part of `TestingStuff ForTestingPurposesOnly()?

Reply:
Great idea